### ISSUE TITLE
Does this repo include energy-based detection with margin?

### ISSUE THREAD

**Issue by amztc34283** [kkirchheim/pytorch-ood#72]
https://www.youtube.com/watch?v=Ga09-9JItxs

I saw the repo have energy-based implementation but there was a new paper using additional margin, do you have this implemented?

If not, I am planning to create a PR.

**Comment by kkirchheim** [comment]
Is this the supervised "Energy Regularization" loss where they enforce an energy gap between IN and OOD data during training? This method was, iirc, proposed in the original energy-based OOD detection paper. If so, it it implemented [here](https://pytorch-ood.readthedocs.io/en/latest/losses.html#energy-bounded-learning-loss). Pretrained ResNet-40 weights from the original repo are also [available](https://pytorch-ood.readthedocs.io/en/latest/models.html#wide-resnet) for CIFAR10+100.

**Comment by amztc34283** [comment]
I don't know if that is the case, this is the link to the paper - [Feed Two Birds with One Scone: Exploiting Wild Data for Both Out-of-Distribution Generalization and Detection](https://proceedings.mlr.press/v202/bai23a/bai23a.pdf). If the paper you are referring to is Energy-based out-of-distribution detection (2020), then the paper I mentioned is an extension of that to handle covariate-shifted data.

Looks like the paper you are referring is the paper from 2020, then I think we could extend to this new method.

**Comment by kkirchheim** [comment]
Okay, I see. No, this is not yet included. If you would like to contribute an implementation, we would highly appreciate it.

**Comment by amztc34283** [comment]
How long does the pytest usually take?

Is there a way to run those tests on my MacBook? Maybe I can't if the fine-tuning takes a long time.

**Comment by kkirchheim** [comment]
The tests should only take a couple of minutes. There is no fine-tuning done, so that should not be an issue. However, some of the tests require network access to download datasets, so if your connection is slow, this might be an issue. Overall, tests should be platform-independent and do not require cuda. Can you narrow it down to a particular test execution?

**Comment by amztc34283** [comment]
Let me get back to you this weekend.

**Comment by amztc34283** [comment]
Since the energy-based OOD with margin requires fine-tuning on pre-trained model, do you want me to write tests that involves fine-tuning? In general, what tests do you suggest to write for a new detection method?

**Comment by kkirchheim** [comment]
> do you want me to write tests that involves fine-tuning?

No, but the tests should ensure that you can do a forward and a backward pass through the method. Maybe have a look at the existing tests to get some inspiration on what to test.

**Comment by amztc34283** [comment]
My repo is almost ready for pull request. I still have to work on some documentation this week and aim to merge by the end of week.

**Comment by amztc34283** [comment]
Closing as merge request is created!
